### PR TITLE
summary page enhancement with bugs ratio added.

### PR DESCRIPTION
**Pull request by rainflood**
In summary page, added two columns in by_project, by_enum, by_developer and by_reporter summary tables: resolved bug ratio, i.e. (resolved + closed)/total *100%, and item ratio, i.e. (item total)/(sum of total bugs).

Fixed #0023627: Summary page enhancement with bugs ratio support

### CODE REVIEW

**Comment by atrol** [comment]
Thanks for your contribution.

Please open a request at https://www.mantisbt.org/bugs for it and reword your commit messages by adding `Fixes #<IdOfYourIssue>` to the message body.

Do not change any other message file than English.
Translations to other languages are managed via Translatewiki, see http://www.mantisbt.org/wiki/doku.php/mantisbt:translationshowto for more details.

I did not have a deeper look at your code changes, especially as I got various system notices when trying to run your branch.

```
SYSTEM NOTICE: 'Undefined index: resolved' in '/Applications/MAMP/htdocs/atrol/mantisbt.org/core/summary_api.php' line 165
```

**Comment by rainflood** [comment]
Thanks for your reply and guidance. I am not familiar with the detail of contribution for mantisbt. I used mantisbt in our devops flow with multi-projects, it's very helpfull. But in manager's view, I need more statistical data, so  I modified summary page code to fulfill the need. 
I searched and read the developer docs for contribution, and submit the codes via pr. I am not sure that should I open a request as 'bug' first even it's a function enhancement, then fix it with pr?

Thanks for the guidance of translation workflow, that wired me for a time.
I will add isset check in that codes.
Thanks again.

**Comment by rainflood** [comment]
Modified commit message with bugid attached and fixed "Undefined index" code.

File: summary_page.php [hunk 1]
```diff
@@ -146,7 +146,7 @@
 <!-- LEFT COLUMN -->
 <div class="col-md-6 col-xs-12">
 
-	<?php if( 1 < count( $t_project_ids ) ) { ?>
+	<?php if( 0 < count( $t_project_ids ) ) { ?>
```
Comment (dregad) [inline summary_page.php]: Count can't be less than zero, so might as well remove the test (and the corresponding closing brace)

File: core/summary_api.php [hunk 2]
```diff
@@ -116,126 +121,98 @@ function summary_print_by_enum( $p_enum ) {
 				ORDER BY ' . $p_enum . ' ' . $t_status_query;
 	$t_result = db_query( $t_query );
 
-	$t_last_value = -1;
+	$t_resolved_val = config_get( 'bug_resolved_status_threshold' );
+	$t_closed_val = config_get( 'bug_closed_status_threshold' );
+
+	$p_cache = array();
```
Comment (dregad) [inline core/summary_api.php]: `p_` prefix is for function parameters, I guess you meant `t_cache` ?

File: core/summary_api.php [hunk 3]
```diff
@@ -116,126 +121,98 @@ function summary_print_by_enum( $p_enum ) {
...
 		# Build up the hyperlinks to bug views
 		$t_bug_link = '';
+		$t_bugs_open = isset($t_item['open']) ? $t_item['open'] : 0;
+		$t_bugs_resolved = isset($t_item['resolved']) ? $t_item['resolved'] : 0;
+		$t_bugs_closed = isset($t_item['closed']) ? $t_item['closed'] : 0;
+		$t_bugs_total = $t_bugs_open + $t_bugs_resolved + $t_bugs_closed;
+		$t_bugs_resolved_ratio = ($t_bugs_resolved + $t_bugs_closed)/($t_bugs_total==0 ? 1 : $t_bugs_total);
+		$t_bugs_ratio = $t_bugs_total /( $t_bugs_total_count == 0 ? 1 : $t_bugs_total_count);
+		$p_bugs_resolved_ratio = sprintf("%.1f%%", $t_bugs_resolved_ratio*100);
+		$p_bugs_ratio = sprintf("%.1f%%", $t_bugs_ratio*100);
```
Comment (dregad) [inline core/summary_api.php]: Coding guidelines: spaces within parentheses.

File: core/summary_api.php [hunk 4]
```diff
@@ -466,68 +443,61 @@ function summary_print_by_developer() {
...
-	$t_bugs_resolved = 0;
-	$t_bugs_closed = 0;
-	$t_bugs_total = 0;
-
 	$t_resolved_val = config_get( 'bug_resolved_status_threshold' );
 	$t_closed_val = config_get( 'bug_closed_status_threshold' );
 
 	$t_summaryusers = array();
-	$t_summarydata = array();
+	$p_cache = array();
```
Comment (dregad) [inline core/summary_api.php]: Same comment as before

File: core/summary_api.php [hunk 5]
```diff
@@ -466,68 +443,61 @@ function summary_print_by_developer() {
...
+			}
 		}
-		$t_last_handler = $v_handler_id;
 	}
+	
+	user_cache_array_rows( array_unique( $t_summaryusers ) );
 
-	if( 0 < $t_bugs_total ) {
-		$t_user = summary_helper_get_developer_label( $t_last_handler );
+	foreach( $p_cache as $t_label=>$t_item) {
```
Comment (dregad) [inline core/summary_api.php]: Spaces around `=>`

File: summary_page.php (hunk 1 above)
Comment (rainflood) [reply summary_page.php]: Count maybe zero, we'd better not print a lonely table header :)

File: core/summary_api.php (hunk 2 above)
Comment (rainflood) [reply core/summary_api.php]: got it.

File: core/summary_api.php (hunk 3 above)
Comment (rainflood) [reply core/summary_api.php]: OK

File: core/summary_api.php (hunk 4 above)
Comment (rainflood) [reply core/summary_api.php]: OK

File: core/summary_api.php (hunk 5 above)
Comment (rainflood) [reply core/summary_api.php]: OK

**Comment by rainflood** [comment]
dregad , thanks for your comments.  Codes modified according Coding guidelines .  I refactored many codes during the enhancement,  so it seems need  further refactor such as to  extract another helper function. Your suggest is very good,  that need dig deep into the code and more time to test, can we make it works better and looks better than before just now,  and do the further refactor later?

File: summary_page.php (hunk 1 above)
Comment (dregad) [reply summary_page.php]: Duh. Just ignore me :wink:

**Comment by dregad** [comment]
> can we make it works better and looks better than before just now, and do the further refactor later?

If you don't mind, I'd rather you take the opportunity to do this refactoring now "while it's hot".

**Comment by rainflood** [comment]
dregad, well, I will try my best to refactor the codes.

File: core/summary_api.php [hunk 6]
```diff
@@ -53,23 +53,27 @@
...
- * Print row in summary table
+ * Print row with percentage in summary table
  *
  * @param string $p_label    The summary row label.
  * @param string $p_open     Count of open issues - normally string with hyperlink to filter.
  * @param string $p_resolved Count of resolved issues - normally string with hyperlink to filter.
  * @param string $p_closed   Count of closed issues - normally string with hyperlink to filter.
  * @param string $p_total    Count of total issues - normally string with hyperlink to filter.
+ * @param string $p_resolved_ratio  ratio of resolved
+ * @param string $p_closed_ratio   ratio of closed
```
Comment (atrol) [inline core/summary_api.php]: The arameter name is $p_ratio

File: core/summary_api.php [hunk 7]
```diff
@@ -92,6 +96,85 @@ function summary_helper_get_developer_label ( $p_user_id ) {
...
+		$p_bugs_total = $p_bug_link . '&amp;' . FILTER_PROPERTY_HIDE_STATUS . '=' . META_FILTER_NONE . '">' . $p_bugs_total . '</a>';
+	}	
+}
+
+/**
+ * Calcute bug ratio 
+ * @param string $p_bugs_open 		The open bugs count.
+ * @param string $p_bugs_resolved 	The resovled bugs count.
+ * @param string $p_bugs_closed   	The closed bugs count.
+ * @param string $p_bugs_total_count   	The total bugs count.
```
Comment (atrol) [inline core/summary_api.php]: All parameters are of type integer

File: core/summary_api.php [hunk 8]
```diff
@@ -92,6 +96,85 @@ function summary_helper_get_developer_label ( $p_user_id ) {
 }
 
 /**
+ * Calculate bug status count according to  'open', 'resolved' and 'closed', 
+ * then put the numbers into $p_cache array
+ *
+ * @param string &$p_cache    The cache array.
```
Comment (atrol) [inline core/summary_api.php]: String? Isn't it array?

File: core/summary_api.php (hunk 6 above)
Comment (rainflood) [reply core/summary_api.php]: oh, yes

File: core/summary_api.php (hunk 7 above)
Comment (rainflood) [reply core/summary_api.php]: changed to integer

File: core/summary_api.php (hunk 8 above)
Comment (rainflood) [reply core/summary_api.php]: array:)

**Comment by rainflood** [comment]
dregad and atrol,  codes has been committed, if there are any other improvements needed,  just tell me.

File: lang/strings_english.txt [hunk 9]
```diff
@@ -1192,7 +1192,7 @@ $s_average_time = 'Average time';
 $s_total_time = 'Total time';
 $s_developer_stats = 'Developer Stats';
 $s_reporter_stats = 'Reporter Stats';
-$s_orct = '(open/resolved/closed/total)';
+$s_orct = '(open/resolved/closed/total/resolved ratio/ratio)';
```
Comment (vboctor) [inline lang/strings_english.txt]: Will this break for non-translated languages?  Should we create a new string and delete this one?

File: core/summary_api.php [hunk 10]
```diff
@@ -92,6 +96,85 @@ function summary_helper_get_developer_label ( $p_user_id ) {
...
+		} else {
+			$p_cache[$p_key]['open'] = $p_bugcount;
+		}
+	}
+}
+/** 
+ * Build bug links for 'open', 'resolved' and 'closed' issue counts
+ * 
+ * @param string $p_bug_link 		The base bug link.
+ * @param string &$p_bugs_open 		The open bugs count, return open bugs link.
```
Comment (vboctor) [inline core/summary_api.php]: Adjust indentation, use spaces between parameter name and description.

File: core/summary_api.php [hunk 11]
```diff
@@ -92,6 +96,85 @@ function summary_helper_get_developer_label ( $p_user_id ) {
...
+		$p_bugs_closed = $p_bug_link . '&amp;' . FILTER_PROPERTY_STATUS . '=' . $t_closed_val . '&amp;' . FILTER_PROPERTY_HIDE_STATUS . '=' . META_FILTER_NONE . '">' . $p_bugs_closed . '</a>';
+	}
+	if( 0 < $p_bugs_total ) {
+		$p_bugs_total = $p_bug_link . '&amp;' . FILTER_PROPERTY_HIDE_STATUS . '=' . META_FILTER_NONE . '">' . $p_bugs_total . '</a>';
+	}	
+}
+
+/**
+ * Calculate bug ratio 
+ * @param integer $p_bugs_open 		The open bugs count.
```
Comment (vboctor) [inline core/summary_api.php]: Fix indentation of parameter description.

File: core/summary_api.php (hunk 10 above)
Comment (rainflood) [reply core/summary_api.php]: Indentation fixed.

File: core/summary_api.php (hunk 11 above)
Comment (rainflood) [reply core/summary_api.php]: Indentation fixed.

File: lang/strings_english.txt (hunk 9 above)
Comment (rainflood) [reply lang/strings_english.txt]: Indentation fixed.  I am not sure about the translation process. While using the old string name, in the non-translated languages the last two new headers in summary tables will not be shown , but the two new columns values will be shown.  If  we create a new string, no headers will be shown in non-translated languages at all.  Please make a decision, after than I will adjust the string file and codes.

Comment (rainflood) [reply lang/strings_english.txt]: Correct: If we create a new string, only English strings headers will be shown in non-translated languages.

Comment (rainflood) [reply lang/strings_english.txt]: Rename the string variable and keep the old one maybe is the better way. I searched the code tree and found   MantisGraph referenced $s_orct  string at line 130, but not used that actually in other codes.
./plugins/MantisGraph/pages/issues_trend_bystatus_table.php:$t_label_string = lang_get( 'orct' ); # use the (open/resolved/closed/total) label

**Comment by rainflood** [comment]
Used a new string name $s_summary_head while keeping the old one $s_orct for compatibility, and rebase the commits onto the master branch.

File: summary_page.php (hunk 1 above)
Comment (dregad) [reply summary_page.php]: Actually, DON'T ignore me... This code is testing whether there we are processing more than one project, so your change is introducing a regression.

The _By Project_ table should only be displayed when viewing the page for _All Projects_. With your change, the table is printed also when a single project is selected (and shows only one line).

File: lang/strings_english.txt [hunk 12]
```diff
@@ -1193,6 +1193,7 @@ $s_total_time = 'Total time';
 $s_developer_stats = 'Developer Stats';
 $s_reporter_stats = 'Reporter Stats';
 $s_orct = '(open/resolved/closed/total)';
+$s_summary_header = '(open/resolved/closed/total/resolved ratio/ratio)';
```
Comment (dregad) [inline lang/strings_english.txt]: Since we're introducing a new language string, I see no reason to maintain the legacy format with the surrounding parentheses, which makes processing the string more complex than it needs to be (preg_split with PREG_SPLIT_NO_EMPTY flag, or explode + substr like in the graph API).

I suggest using a simple delimited string, so a simple _explode()_ call can do the job.

File: summary_page.php [hunk 13]
```diff
@@ -111,13 +111,13 @@
 $t_total_time		= number_format( $t_total_time / SECONDS_PER_DAY, 2 );
 $t_average_time 	= number_format( $t_average_time / SECONDS_PER_DAY, 2 );
 
-$t_orct_arr = preg_split( '/[\)\/\(]/', lang_get( 'orct' ), -1, PREG_SPLIT_NO_EMPTY );
+$t_summary_header_arr = preg_split( '/[\)\/\(]/', lang_get( 'summary_header' ), -1, PREG_SPLIT_NO_EMPTY );
```
Comment (dregad) [inline summary_page.php]: Use a simple _explode()_ call, see earlier comment

Comment (rainflood) [reply summary_page.php]: Refactored string split with explode(), feels digging deep...

About "By Project"  table, we want to see the bug ratios of the only one project selected in the whole project view, also there maybe subprojects as well. Perhaps we can change the header title form "By Project" to "Project" which string name I find is "$s_email_project" in the language file, when only one project selected. 
$ grep \'Project\' lang/strings_english.txt 
$s_email_project = 'Project';
Codes in summary_page.php like this:
@@ -153,7 +153,11 @@ print_summary_submenu();
                <table class="table table-hover table-bordered table-condensed table-striped">
                <thead>
                        <tr>
+                       <?php if( 1 == count( $t_project_ids ) ) { ?>
+                               <th class="width-35"><?php echo lang_get( 'email_project' ) ?></th>
+                       <?php } else { ?>
                                <th class="width-35"><?php echo lang_get( 'by_project' ) ?></th>
+                       <?php } ?>
                                <?php echo $t_summary_header ?>
                        </tr>
                </thead>

File: summary_page.php [hunk 14]
```diff
@@ -153,7 +153,11 @@
 		<table class="table table-hover table-bordered table-condensed table-striped">
 		<thead>
 			<tr>
+			<?php if( 1 == count( $t_project_ids ) ) { ?>
+				<th class="width-35"><?php echo lang_get( 'email_project' ) ?></th>
```
Comment (atrol) [inline summary_page.php]: It's confusing to use `email_project` for something that is not related to email.
I would introduce a new string for it.

File: summary_page.php (hunk 1 above)
Comment (dregad) [reply summary_page.php]: Re-posting @rainflood's note from https://github.com/mantisbt/mantisbt/pull/1231#discussion_r161386203 to follow a single conversation flow.

About "_By Project_"  table, we want to see the bug ratios of the only one project selected in the whole project view, also there maybe subprojects as well. Perhaps we can change the header title form "_By Project_" to "_Project_" which string name I find is "$s_email_project" in the language file, when only one project selected. 
$ grep \'Project\' lang/strings_english.txt 
$s_email_project = 'Project';
Codes in summary_page.php like this:

```php
@@ -153,7 +153,11 @@ print_summary_submenu();
                <table class="table table-hover table-bordered table-condensed table-striped">
                <thead>
                        <tr>
+                       <?php if( 1 == count( $t_project_ids ) ) { ?>
+                               <th class="width-35"><?php echo lang_get( 'email_project' ) ?></th>
+                       <?php } else { ?>
                                <th class="width-35"><?php echo lang_get( 'by_project' ) ?></th>
+                       <?php } ?>
                                <?php echo $t_summary_header ?>
                        </tr>
                </thead>
```

Comment (dregad) [reply summary_page.php]: The case of subprojects is already covered by the current code; the box is only hidden when showing summary for a single project. I am not opposed to having it displayed at all times, but that change was not documented as such, so I considered it as not in the PR's scope.

Quite frankly I do not see the added value of a differentiated title depending on the number of projects, it feels like needlessly complicating things. Moving forward, I would propose that

1. the change is properly documented (i.e. create an issue in our tracker to describe it)
2. reference the new issue in the message of commit 72be8c709c271d39481f3f01bc25bb6949fc582a
3. revert 2a01f01003164e81e8cb8ef8e94d9496f8890797

File: core/summary_api.php [hunk 15]
```diff
@@ -53,23 +53,27 @@
...
+function summary_helper_print_row( $p_label, $p_open, $p_resolved, $p_closed, $p_total, $p_resolved_ratio, $p_ratio) {
+        echo '<tr>';
+        printf( '<td class="width50">%s</td>', $p_label );
+        printf( '<td class="width12 align-right">%s</td>', $p_open );
+        printf( '<td class="width12 align-right">%s</td>', $p_resolved );
+        printf( '<td class="width12 align-right">%s</td>', $p_closed );
+        printf( '<td class="width12 align-right">%s</td>', $p_total );
+        printf( '<td class="width12 align-right">%s</td>', $p_resolved_ratio );
+        printf( '<td class="width12 align-right">%s</td>', $p_ratio );
+        echo '</tr>';
```
Comment (dregad) [inline core/summary_api.php]: Wrong indentation type - should be a tab, not spaces

File: core/summary_api.php [hunk 16]
```diff
@@ -116,126 +199,76 @@ function summary_print_by_enum( $p_enum ) {
 				ORDER BY ' . $p_enum . ' ' . $t_status_query;
 	$t_result = db_query( $t_query );
 
-	$t_last_value = -1;
+	$t_cache = array();
+	$t_bugs_total_count = 0;
 	$t_bugs_open = 0;
 	$t_bugs_resolved = 0;
 	$t_bugs_closed = 0;
 	$t_bugs_total = 0;
```
Comment (dregad) [inline core/summary_api.php]: The above 4 assignments are unnecessary, the variables' values are overwritten in the foreach loop

File: core/summary_api.php [hunk 17]
```diff
@@ -91,6 +95,85 @@ function summary_helper_get_developer_label ( $p_user_id ) {
...
+	}
+}
+/** 
+ * Build bug links for 'open', 'resolved' and 'closed' issue counts
+ * 
+ * @param string $p_bug_link            The base bug link.
+ * @param string &$p_bugs_open          The open bugs count, return open bugs link.
+ * @param string &$p_bugs_resolved      The resovled bugs count, return resolved bugs link.
+ * @param string &$p_bugs_closed        The closed bugs count, return closed bugs link.
+ * @param string &$p_bugs_total         The total bugs count, return total bugs link.
```
Comment (dregad) [inline core/summary_api.php]: I think those 4 params are integers, not strings

File: core/summary_api.php [hunk 18]
```diff
@@ -466,82 +499,38 @@ function summary_print_by_developer() {
...
 	$t_result = db_query( $t_query );
 
-	$t_last_handler = -1;
+	$t_summaryusers = array();
+	$t_cache = array();
+	$t_bugs_total_count = 0;
 	$t_bugs_open = 0;
 	$t_bugs_resolved = 0;
 	$t_bugs_closed = 0;
-	$t_bugs_total = 0;
```
Comment (dregad) [inline core/summary_api.php]: The above 4 assignments are unnecessary, the variables' values are overwritten in the foreach loop

File: core/summary_api.php [hunk 19]
```diff
@@ -466,82 +499,38 @@ function summary_print_by_developer() {
...
-		}
-		if( 0 < $t_bugs_closed ) {
-			$t_bugs_closed = $t_bug_link . '&amp;' . FILTER_PROPERTY_STATUS . '=' . $t_closed_val . '&amp;' . FILTER_PROPERTY_HIDE_STATUS . '=' . META_FILTER_NONE . '">' . $t_bugs_closed . '</a>';
-		}
-		if( 0 < $t_bugs_total ) {
-			$t_bugs_total = $t_bug_link . '&amp;' . FILTER_PROPERTY_HIDE_STATUS . '=' . META_FILTER_NONE . '">' . $t_bugs_total . '</a>';
-		}
+	foreach( $t_cache as $t_label => $t_item) {
+		# Build up the hyperlinks to bug views
+		$t_bug_link = '';
```
Comment (dregad) [inline core/summary_api.php]: Useless assignment, $t_bug_link is initialized further down

File: core/summary_api.php [hunk 20]
```diff
@@ -649,86 +643,37 @@ function summary_print_by_category() {
 
 	$t_result = db_query( $t_query );
 
-	$t_last_category_name = -1;
-	$t_last_project = -1;
+	$t_cache = array();
+	$t_category_name = "";
+	$t_project = "";
```
Comment (dregad) [inline core/summary_api.php]: $t_category_name and $t_project are not used

File: core/summary_api.php [hunk 21]
```diff
@@ -649,86 +643,37 @@ function summary_print_by_category() {
...
+	$t_bugs_total_count = 0;
 	$t_bugs_open = 0;
 	$t_bugs_resolved = 0;
 	$t_bugs_closed = 0;
```
Comment (dregad) [inline core/summary_api.php]: The above 3 assignments are unnecessary, the variables' values are overwritten in the foreach loop

File: summary_page.php (hunk 1 above)
Comment (dregad) [reply summary_page.php]: > 1. the change is properly documented (i.e. create an issue in our tracker to describe it)

I just opened [#23854](https://mantisbt.org/bugs/view.php?id=23854)

File: core/summary_api.php (hunk 17 above)
Comment (dregad) [reply core/summary_api.php]: Nevermind my comment - it's _integer_ input and _string_ output, so might as well leave the PHPDoc as-is.

Comment (rainflood) [reply core/summary_api.php]: dregad, sorry for respond late. After the crazy busy weekdays, I saw you have fix the requested, are there anymore need me to commit?

**Comment by dregad** [comment]
In response to @rainflood https://github.com/mantisbt/mantisbt/pull/1231#discussion_r162771718

> dregad, sorry for respond late. After the crazy busy weekdays, I saw you have fixed the request, are there anymore need me to commit?

Yes I had a bit of spare time, and needed this done to continue work on other issues with summary page. I don't think further input is required from your end at this time. Thank you.

**Comment by dregad** [comment]
Merged in 
- 1c2b9340f29c8878a6a3a52d05ec5d96f7c83f02 (add ratios)
- b6eeaa017e3a14391312f8ff7e36ed246a739e0b (always display "by project)
